File: src/Niche/Nim/prototype.nim
import std/[sequtils, strutils]

type
  ServiceProfile = object
    name: string
    features: seq[string]

proc cloneProfile(profile: ServiceProfile): ServiceProfile =
  ServiceProfile(name: profile.name, features: profile.features.mapIt(it))

proc describe(profile: ServiceProfile): string =
  profile.name & ": " & profile.features.join(",")

var original = ServiceProfile(name: "orders", features: @["metrics"])
var canary = cloneProfile(original)
canary.name = "orders-canary"
canary.features.add("tracing")

echo "original=" & describe(original)
echo "clone=" & describe(canary)

File: src/Web/Dart/prototype.dart
class ServiceProfile {
  ServiceProfile(this.name, List<String> features)
    : features = List<String>.of(features);

  String name;
  final List<String> features;

  ServiceProfile cloneProfile() => ServiceProfile(name, features);

  String describe() => '$name: ${features.join(',')}';
}

void main() {
  final original = ServiceProfile('orders', ['metrics']);
  final canary = original.cloneProfile()
    ..name = 'orders-canary'
    ..features.add('tracing');

  print('original=${original.describe()}');
  print('clone=${canary.describe()}');
}
